### ISSUE TITLE
COM Collector needs to handle indexed TLBs and referenced, but unregistered type libraries

### ISSUE THREAD

**Issue by ThunderFrame** [rubberduck-vba/Rubberduck#4193]
Using a command like `ThisWorkbook.VBProject.References.AddFromFile "C:\Program Files (x86)\Common Files\microsoft shared\VBA\VBA7.1\VBE7.dll\3"` (note the index numeral, at the tail of the path), it's possible to add a `Reference` that isn't the default TLB for a file (in this case, the `VBInternal` TLB is referenced), and that reference need not be registered. VBA happily allows this, but Rubberduck is unable to successfully parse the type library.

The reasons are potentially twofold:
- `VBInternal` is not registered, so a call to `References("VBInternal").FullPath` results in an error. 
    > Run-time error '-2147319779 (8002801d)': 
    > Automation error
    > Library not registered. 
- The COM Collector is unaware of the indexed path syntax for targeting a particular TLB within a file

Interestingly, the References dialog does reveal the reference's complete path and index, as shown below, so it might be that VBIDE won't reveal the path, but the ITypeLib might.
![vbint](https://user-images.githubusercontent.com/16574009/42758841-354a4ffc-8948-11e8-9a6b-781f0d74e16b.png)

Also, the Reference object also has a `Guid` property, and `VBInternal` does have a Guid, so it might be possible to discover the path, via the registry, to the file.

Once the COM collector has the path, we'll need to experiment with how to reach the indexed Type Library.

Once the COM Collector can find the TLB, it should be able to parse the TLB as if it was in a normal path.

**Comment by comintern** [comment]
The more I dig into this, the more I suspect that it's a special case with the `VBInternal` typelib. We actually have a serialized library that uses an indexed type library (vbscript.dll is on index 3), and that one loads just fine when it's passed into LoadTypeLibEx. This seems to agree the documentation for [LoadTypeLib](https://docs.microsoft.com/en-gb/previous-versions/windows/desktop/api/oleauto/nf-oleauto-loadtypelib), which states:

> If the file is a DLL or an executable file, it is loaded. By default, the type library is extracted from the first resource of type ITypeLib. To load a different type of library resource, append an integer index to szFile.

So, to paraphrase @ThunderFrame, the reason is potentailly onefold - VBE's `References().FullPath` implementation is jacked up in this case because RD's current loading implementation should "Just Work™". I may need to investigate whether any oleaut functions are "aware" of unregistered type libraries that have already been loaded in-process. If not, it might be difficult to associate the GUID from the IReference to a file it can be manually loaded from.

If push comes to shove and this is a corner case due to the VBE being stupid, we can just special case the GUID for `VBInternal`. Do we know if the VBE registers added references from the browse dialog if they aren't already in the registry? I'm also curious if any other references exhibit the same behavior.

**Comment by bclothier** [comment]
FWIW.... the typelib API can provide us with the information. This is the output from the `DocumentAll`:

```
- VBE Reference: Visual Basic For Applications [path: C:\PROGRA~2\COMMON~1\MICROS~1\VBA\VBA7\VBE7.DLL, majorVersion: 4, minorVersion: 1, guid: 000204ef-0000-0000-c000-000000000046, lcid: 9]
- VBE Reference: Microsoft Excel 14.0 Object Library [path: C:\Program Files (x86)\Microsoft Office\Office14\EXCEL.EXE, majorVersion: 1, minorVersion: 7, guid: 00020813-0000-0000-c000-000000000046, lcid: 0]
- VBE Reference: OLE Automation [path: C:\Windows\SysWOW64\stdole2.tlb, majorVersion: 2, minorVersion: 0, guid: 00020430-0000-0000-c000-000000000046, lcid: 0]
- VBE Reference: Microsoft Office 14.0 Object Library [path: C:\Program Files (x86)\Common Files\Microsoft Shared\OFFICE14\MSO.DLL, majorVersion: 2, minorVersion: 5, guid: 2df8d04c-5bfa-101b-bde5-00aa0044de52, lcid: 0]
- VBE Reference: VBInternal [path: C:\Program Files (x86)\Common Files\microsoft shared\VBA\VBA7.1\VBE7.dll\3, majorVersion: 6, minorVersion: 0, guid: 6715f050-f076-11cd-9b33-08002b336f14, lcid: 9]
- VBE Reference: Microsoft Forms 2.0 Object Library [path: C:\Windows\SysWOW64\FM20.DLL, majorVersion: 2, minorVersion: 0, guid: 0d452ee1-e08f-101a-852e-02608c4d0bb4, lcid: 0]
- VBE Reference: Microsoft VBScript Regular Expressions 5.5 [path: C:\Windows\System32\vbscript.dll\3, majorVersion: 5, minorVersion: 5, guid: 3f4daca7-160d-11d2-a8e9-00104b365c9f, lcid: 0]
- VBE Reference: Microsoft VBScript Globals [path: C:\Windows\System32\vbscript.dll, majorVersion: 1, minorVersion: 0, guid: 3eef9758-35fc-11d1-8ce4-00c04fc2b185, lcid: 0]
- VBE Reference: Microsoft VBScript Regular Expressions 1.0 [path: C:\Windows\System32\vbscript.dll\2, majorVersion: 1, minorVersion: 0, guid: 3f4daca7-160d-11d2-a8e9-00104b365c9f, lcid: 0]
```

Therefore, if it's referenced in the VBA project, you could then leverage the typelib API and thus bypass the `LoadTypeLib`.

**Comment by WaynePhillipsEA** [comment]
@comintern the TypeLib Api exposes the loaded references for each project, so in theory you can grab the ITypelib for each reference already loaded by the VBE

**Comment by bclothier** [comment]
@WaynePhillipsEA to be fair, you shot from hip and hit. I had to fumble with the gun a bit first. :) 

@comintern  regarding this bit:

> Do we know if the VBE registers added references from the browse dialog if they aren't already in the registry? I'm also curious if any other references exhibit the same behavior.

The library need not be listed to be referenced. I can browse to any tlb file and it can be added to the reference dialog. However, I don't think that is usable because VBA need to be able to locate the DLL that provides the implementation, and I _think_ it uses the `CLSID`/`Interface` registry records to do so; so the resulting code will compile but won't run. In another project, that tlb file I added won't be listed and won't stay listed until I actually register them in the registry.
